src/notation/view/internal/dynamicpopupmodel.cpp

```diff
@@ -0,0 +1,225 @@
+
+    for (const QList<DynamicPopupModel::PageItem>& page : DYN_POPUP_PAGES) {
+        QVariantList variantPage;
+        for (const DynamicPopupModel::PageItem& item : page) {
+            QVariantMap variantMap {
+                { "text", xmlTextToQString(item.text, engravingFont) },
+                { "width", item.width },
+                { "offset", item.offset },
+                { "type", item.itemType },
+            };
+            variantPage.append(variantMap);
+        }
+        m_pages.append(QVariant::fromValue(variantPage));
+    }
+
+    emit pagesChanged();
+}
+
+void DynamicPopupModel::addOrChangeDynamic(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
+        return;
+    }
+
+    beginCommand();
+    m_item->undoChangeProperty(Pid::TEXT, Dynamic::dynamicText(DYN_POPUP_PAGES[page][index].dynType));
+    m_item->undoChangeProperty(Pid::DYNAMIC_TYPE, DYN_POPUP_PAGES[page][index].dynType);
+    endCommand();
+
+    if (m_item->flag(ElementFlag::IS_PREVIEW)) {
+        m_item->setFlag(ElementFlag::IS_PREVIEW, false);
+    }
+
+    INotationInteractionPtr interaction = currentNotation()->interaction();
+
+    interaction->flipHairpinsType(toDynamic(m_item));
+
+    // Hide the bounding box which appears when called using Ctrl+D shortcut
+    if (interaction->isTextEditingStarted()) {
+        interaction->endEditText();
+        interaction->startEditGrip(m_item, Grip::DRAG);
+    }
+
+    updateNotation();
+}
+
+void DynamicPopupModel::addHairpinToDynamic(ItemType itemType)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
+        return;
+    }
+
+    beginCommand();
+
+    Hairpin* pin = Factory::createHairpin(m_item->score()->dummy()->segment());
+    if (itemType == ItemType::Crescendo) {
+        pin->setHairpinType(HairpinType::CRESC_HAIRPIN);
+    } else if (itemType == ItemType::Decrescendo) {
+        pin->setHairpinType(HairpinType::DECRESC_HAIRPIN);
+    }
+
+    m_item->score()->addHairpinToDynamic(pin, toDynamic(m_item));
+
+    endCommand();
+    updateNotation();
+}
+
+void DynamicPopupModel::showPreview(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic() || !m_item->flag(ElementFlag::IS_PREVIEW)) {
+        return;
+    }
+
+    beginCommand();
+    m_item->setProperty(Pid::TEXT, Dynamic::dynamicText(DYN_POPUP_PAGES[page][index].dynType));
+    m_item->setProperty(Pid::DYNAMIC_TYPE, DYN_POPUP_PAGES[page][index].dynType);
+    endCommand();
+    updateNotation();
+}
+
+void DynamicPopupModel::hidePreview()
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic() || !m_item->flag(ElementFlag::IS_PREVIEW)) {
+        return;
+    }
+
+    beginCommand();
+    m_item->setProperty(Pid::TEXT, String());
+    m_item->setProperty(Pid::DYNAMIC_TYPE, DynamicType::OTHER);
+    endCommand();
```

Comment:
I'm afraid this is problematic: it means that these changes will be stored in the undo stack. Suppose the user does the following:
1. Create dynamic
2. Activate preview by hovering
3. Deactivate preview by un-hovering
4. Press undo

Expected behaviour would be that step 4 brings you back to the state before step 1, but instead, it brings you back to the state before step 3, so with the preview dynamic visible. And this time, without the preview flag (because it goes via the Undo stack, and that doesn't set the preview flag). 

A way to fix this to some extent would be to make the "hidePreview" method call undo. This is still quite fragile though, because you can imagine that all sorts of problems can occur when the user does other things between activating and deactivating the preview. 

This 'preview' functionality is a completely new concept and I think we have to conclude that the current codebase is not quite ready for it. This idea with `ElementFlag::IS_PREVIEW` solves the problem very decently and gracefully, but only if you don't take 'undo' into account. Making it undo-proof seems much more difficult.

When splitting the code into multiple PRs, I would suggest to move the changes about the preview functionality into a separate, fourth PR, because these will simply require a lot more thought, also from us.